### PR TITLE
MQTT Polling Retries & Reliability Updates

### PR DESCRIPTION
**Pull request by mechawrench**
**Description**
- Polls MQTT with up to 10 failures in 30 seconds before displaying the error page to go back and try again
  - This should now cleanly exit without an error message that requires a reboot on MQTT failures
  - I don't anticipate hitting 10, the failures seem to be intermittent and previously a failure for any reason would immediately crash the WiFiCom
  - Also in mqtt.send_digirom_output() we try/except the publish, later plans to add a max failure rate here too
- I reused the code for displaying failure via code.py, so the user can simply go back and try to connect again in the case of a failure

**Bugfixes**
- Pylint inconsistent-return-types fixed in code.py and picow.py

**TODO**
- [x] Test this by simulating a MQTT failure (bring down QA MQTT while connected and watch it fail gracefully)

### ADDED
- MQTT polling retries so a failure doesn't occur immediately on the first intermittent issues, gives MQTT a chance to recover
   - On 10 failures in 30 seconds you will be kicked back to an error screen prompting to hold C to reboot (and can then try again)
   - mqtt.send_digirom_output() is also try/excepted now
 
### Tested with
- CircuitPython 8.0.4

### CODE REVIEW

File: code.py [hunk 1]
```diff
@@ -251,13 +251,45 @@ def run_wifi():
...
+	if loop_result:
+		# If the result is True, continue without incrementing failure_count
+		pass
+	else:
+		# If the result is False, increment failure_count
+		failure_count += 1
+		print(f"Failure count: {failure_count}")
+
+	if time.monotonic() - start_time <= mqtt_loop_timeout:
+		# Quit if 30 seconds have passed since the start time
```
[the remark this answers is not in the document]
Comment (mechawrench) [reply code.py:283]: Fixed this, removed the mqtt_loop_timeout completely and set to a static 30 seconds

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,14 @@ All notable changes to this project will be documented in this file.
...
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+### Added
+- Logic to detect MQTT failures and display error to reboot by holding button C if more than 10 failures occur
+- Logic to detect MQTT Publish failures
+### Tested with
+- CircuitPython 8.0.4
+- MiniMQTT 7.3.2
+
```
[the remark this answers is not in the document]
Comment (mechawrench) [reply CHANGELOG.md]: Moved to PR description

File: code.py [hunk 3]
```diff
@@ -251,13 +251,45 @@ def run_wifi():
...
-				mqtt.loop()
+				mqtt_loop()
 				if mqtt.get_subscribed_output(False) is not None:
 					break
 				if time.monotonic() - time_start >= 5:
 					break
 				time.sleep(0.1)
 
+failure_count = 0
+start_time = time.monotonic()
```
[the remark this answers is not in the document]
Comment (mechawrench) [reply code.py]: That makes sense, went with your suggestion on names

File: lib/wificom/mqtt.py [hunk 4]
```diff
@@ -70,7 +70,12 @@ def loop():
 	'''
 	Loop IO MQTT client
 	'''
-	_mqtt_client.loop()
+	try:
+		_mqtt_client.loop()
+		return True
+	# pylint: disable=broad-except
+	except Exception:
+		return False
```
[the remark this answers is not in the document]
Comment (mechawrench) [reply lib/wificom/mqtt.py:78]: Giving {repr(e)} a try with this PR

I think I have the pylint disable proper now on both of the broad-excepts that are new to this PR now.  Let me know if I understood proper in the implementation

File: lib/wificom/mqtt.py [hunk 5]
```diff
@@ -11,6 +11,7 @@
 secrets_device_uuid, \
 secrets_user_uuid
 
+mqtt_failure_count = 0
```
Comment (dmcomm) [inline lib/wificom/mqtt.py:14]: Since it's in the mqtt module now, could change it back to failure_count.

File: lib/wificom/mqtt.py [hunk 6]
```diff
@@ -30,6 +31,13 @@
 _io = None
 _mqtt_client = None
 
+def get_failure_count():
+	'''
+	Get failure count, used in code.py for catching publish failures
+	'''
+	global mqtt_failure_count # pylint: disable=global-variable-not-assigned
+	return mqtt_failure_count
+
```
Comment (dmcomm) [inline lib/wificom/mqtt.py:40]: We're not using getters in this module.

File: code.py [hunk 7]
```diff
@@ -251,13 +251,43 @@ def run_wifi():
...
+				mqtt_loop()
 				if mqtt.get_subscribed_output(False) is not None:
 					break
 				if time.monotonic() - time_start >= 5:
 					break
 				time.sleep(0.1)
 
+mqtt_failure_start_time = time.monotonic()
+
+def mqtt_loop():
```
Comment (dmcomm) [inline code.py:263]: How about moving all the failure counting into mqtt.loop(), giving that a return value, and just having the final error report here?
```
if not mqtt.loop():
	print("Maximum number of failures reached within 30 seconds. MQTT Failure...")
	connection_failure_alert("MQTT")
```

File: code.py [hunk 8]
```diff
@@ -209,6 +209,7 @@ def run_wifi():
 	led.frequency = 1000
 	led.duty_cycle = LED_DUTY_CYCLE_DIM
 	ui.display_text("WiFi\nHold C to change")
+	mqtt.check_failure_state()
```
Comment (dmcomm) [inline code.py:212]: This one isn't checking the return value. Is it needed?

File: lib/wificom/mqtt.py [hunk 9]
```diff
@@ -120,7 +157,12 @@ def send_rtb_digirom_output(output):
 		mqtt_message_json = json.dumps(mqtt_message)
 
 		if rtb_active:
-			_mqtt_client.publish(rtb_host + '/f/' + rtb_topic, mqtt_message_json)
+			try:
+				_mqtt_client.publish(rtb_host + '/f/' + rtb_topic, mqtt_message_json)
+			except Exception as e: # pylint: disable=broad-except
+				global failure_count
+				failure_count += 1
+				print(f"Failed to connect to send MQTT publish message for RTB: {repr(e)}")  # pylint: disable=broad-except
```
Comment (dmcomm) [inline lib/wificom/mqtt.py:165]: Unnecessary pylint disable

File: code.py [hunk 10]
```diff
@@ -248,16 +251,29 @@ def run_wifi():
...
 			while True:
-				mqtt.loop()
+				mqtt_loop()
 				if mqtt.get_subscribed_output(False) is not None:
 					break
 				if time.monotonic() - time_start >= 5:
 					break
 				time.sleep(0.1)
 
+mqtt_failure_start_time = time.monotonic()
```
Comment (dmcomm) [inline code.py:266]: Don't need this any more.

File: lib/wificom/mqtt.py [hunk 11]
```diff
@@ -70,7 +72,37 @@ def loop():
...
+		_mqtt_client.loop()
+		return True
+	except Exception as e: # pylint: disable=broad-except
+		print(f"Failed to run MQTT loop: {repr(e)}")
+		global failure_count
+		failure_count += 1
+		print(f"Failure count: {failure_count}")
+		return True
+
+def check_failure_state():
```
Comment (dmcomm) [inline lib/wificom/mqtt.py:85]: is_failure_count_ok as discussed

**Comment by dmcomm** [comment]
Actually this could be a class:

`FailureCounter`
variables: `count`, `start_time`
functions: `is_ok:bool`, `report(message:str)`

code.py: `if not mqtt.failure_counter.is_ok()`

**Comment by mechawrench** [comment]
Closed this out, attempted a few ways but we feel this isn't appropriate in the current form.

Maybe will be revisited later

**Pull request closed without merging**